File: TradingFarm/next-dashboard/src/types/trading.ts
/**
 * Trading Farm - Phase 4 Trading Types
 * Type definitions for live trading functionality
 */

// Order types
export enum OrderSide {
  Buy = "buy",
  Sell = "sell"
}

export enum OrderType {
  Market = "market",
  Limit = "limit",
  Stop = "stop",
  StopLimit = "stop_limit"
}

export enum OrderStatus {
  Pending = "pending",
  Open = "open",
  Filled = "filled",
  Partial = "partial",
  Cancelled = "cancelled",
  Rejected = "rejected"
}

export enum TimeInForce {
  GTC = "gtc", // Good Till Cancelled
  IOC = "ioc", // Immediate or Cancel
  FOK = "fok"  // Fill or Kill
}

// Position types
export interface Position {
  id?: number;
  user_id: string;
  agent_id?: string;
  symbol: string;
  quantity: number;
  avg_price: number;
  unrealised_pnl: number;
  created_at?: string;
  updated_at?: string;
}

// Order types
export interface Order {
  id?: number;
  user_id: string;
  agent_id?: string;
  exchange: string;
  symbol: string;
  side: OrderSide;
  type: OrderType;
  price?: number;
  quantity: number;
  status: OrderStatus;
  executed_qty: number;
  tx_id?: string;
  time_in_force?: TimeInForce;
  created_at?: string;
  updated_at?: string;
}

export interface OrderRequest {
  agent_id?: string;
  exchange: string;
  symbol: string;
  side: OrderSide;
  type: OrderType;
  price?: number;
  quantity: number;
  time_in_force?: TimeInForce;
}

export interface OrderResponse {
  id: number;
  exchange: string;
  symbol: string;
  side: string;
  type: string;
  price?: number;
  quantity: number;
  status: string;
  tx_id?: string;
  message?: string;
}

// Risk management types
export interface RiskProfile {
  id?: number;
  user_id: string;
  agent_id?: string;
  max_position_pct: number;
  max_daily_loss: number;
  circuit_breaker: boolean;
  created_at?: string;
  updated_at?: string;
}

// Market data types
export interface MarketData {
  symbol: string;
  exchange: string;
  price: number;
  bid?: number;
  ask?: number;
  volume_24h?: number;
  timestamp: string;
  is_stale: boolean;
}

// Wallet and balance types
export interface WalletBalance {
  id?: number;
  user_id: string;
  exchange: string;
  currency: string;
  free: number;
  locked: number;
  updated_at?: string;
}

// Performance metrics
export interface TradePerformance {
  id?: number;
  user_id: string;
  agent_id?: string;
  period: string;
  pnl: number;
  win_rate?: number;
  sharpe_ratio?: number;
  drawdown?: number;
  created_at?: string;
  updated_at?: string;
}

// API error response
export interface ApiError {
  status: number;
  message: string;
  detail?: string;
}

// Exchange info
export interface Exchange {
  id: string;
  name: string;
  logo?: string;
  features: string[];
  markets: number;
  status: 'online' | 'maintenance' | 'offline';
}

File: TradingFarm/backend/app/models/trading.py
from enum import Enum
from typing import Optional, List, Dict, Any, Union
from datetime import datetime
from pydantic import BaseModel, Field, validator


class OrderSide(str, Enum):
    BUY = "buy"
    SELL = "sell"


class OrderType(str, Enum):
    MARKET = "market"
    LIMIT = "limit"
    STOP = "stop"
    STOP_LIMIT = "stop_limit"


class OrderStatus(str, Enum):
    PENDING = "pending"
    OPEN = "open"
    FILLED = "filled"
    PARTIAL = "partial"
    CANCELLED = "cancelled"
    REJECTED = "rejected"


class TimeInForce(str, Enum):
    GTC = "gtc"  # Good Till Cancelled
    IOC = "ioc"  # Immediate or Cancel
    FOK = "fok"  # Fill or Kill


class Position(BaseModel):
    id: Optional[int] = None
    user_id: str
    agent_id: Optional[str] = None
    symbol: str
    quantity: float
    avg_price: float
    unrealised_pnl: float = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class Order(BaseModel):
    id: Optional[int] = None
    user_id: str
    agent_id: Optional[str] = None
    exchange: str
    symbol: str
    side: OrderSide
    type: OrderType
    price: Optional[float] = None
    quantity: float
    status: OrderStatus = OrderStatus.PENDING
    executed_qty: float = 0
    tx_id: Optional[str] = None
    time_in_force: Optional[TimeInForce] = TimeInForce.GTC
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    @validator('price', pre=True, always=True)
    def validate_price(cls, v, values):
        # Price is required for limit and stop-limit orders
        if values.get('type') in [OrderType.LIMIT, OrderType.STOP_LIMIT] and v is None:
            raise ValueError('Price is required for limit and stop-limit orders')
        return v


class OrderRequest(BaseModel):
    agent_id: Optional[str] = None
    exchange: str
    symbol: str
    side: OrderSide
    type: OrderType
    price: Optional[float] = None
    quantity: float
    time_in_force: Optional[TimeInForce] = TimeInForce.GTC


class OrderResponse(BaseModel):
    id: int
    exchange: str
    symbol: str
    side: str
    type: str
    price: Optional[float] = None
    quantity: float
    status: str
    tx_id: Optional[str] = None
    message: Optional[str] = None


class ExchangeCredentials(BaseModel):
    id: Optional[int] = None
    user_id: str
    exchange: str
    api_key_encrypted: str
    api_secret_encrypted: str
    passphrase: Optional[str] = None
    is_active: bool = True
    last_used: Optional[datetime] = None
    last_failed: Optional[datetime] = None


class WalletBalance(BaseModel):
    id: Optional[int] = None
    user_id: str
    exchange: str
    currency: str
    free: float = 0
    locked: float = 0
    updated_at: Optional[datetime] = None


class MarketData(BaseModel):
    symbol: str
    exchange: str
    price: float
    bid: Optional[float] = None
    ask: Optional[float] = None
    volume_24h: Optional[float] = None
    timestamp: datetime
    is_stale: bool = False


class RiskProfile(BaseModel):
    id: Optional[int] = None
    user_id: str
    agent_id: Optional[str] = None
    max_position_pct: float = 0.05  # Default to 5% of portfolio
    max_daily_loss: float = 0.02    # Default to 2% daily loss limit
    circuit_breaker: bool = False   # Disable circuit breaker by default
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class TradePerformance(BaseModel):
    id: Optional[int] = None
    user_id: str
    agent_id: Optional[str] = None
    period: str  # daily, weekly, monthly
    pnl: float
    win_rate: Optional[float] = None
    sharpe_ratio: Optional[float] = None
    drawdown: Optional[float] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
